plugins/inventory/gql_inventory.py

```diff
@@ -0,0 +1,328 @@
+                        reason=json.loads(e.fp.read().decode())["errors"][0]["message"],
+                    ),
+                    color="red",
+                )
+                # Need to return mock response data that is empty to prevent any failures downstream
+                return {"results": [], "next": None}
+        json_data = json.loads(response.read())
+
+        # Error handling in case of a malformed query
+        if "errors" in json_data:
+            self.display.display(
+                "Query returned an error.\nReason: {0}".format(
+                    json_data["errors"][0]["message"]
+                ),
+                color="red",
+            )
+            # Need to return mock response data that is empty to prevent any failures downstream
+            return {"results": [], "next": None}
+
+        groups = {}
+        if self.group_by:
+            for group_by in self.group_by:
+                if not GROUP_BY.get(group_by):
+                    self.display.display(
+                        "WARNING: '{0}' is not supported as a 'group_by' option. Supported options are: {1} ".format(
+                            group_by,
+                            " ".join("'{0}',".format(str(x)) for x in GROUP_BY.keys()),
+                        ),
+                        color="yellow",
+                    )
+                    continue
+                for device in json_data["data"]["devices"]:
+                    groups[device["site"]["name"]] = "site"
+                    if device.get(group_by) and GROUP_BY.get(group_by):
+                        groups[device[group_by][GROUP_BY.get(group_by)]] = group_by
+                    else:
+                        groups["unknown"] = "unknown"
+
+        else:
+            for device in json_data["data"]["devices"]:
+                groups[device["site"]["name"]] = "site"
+
+        for key, value in groups.items():
+            for device in json_data["data"]["devices"]:
+                if device.get(value) and key == device[value][GROUP_BY[value]]:
+                    self.create_inventory(key, device["name"])
+                    if device["primary_ip4"]:
+                        self.add_variable(
+                            device["name"],
+                            device["primary_ip4"]["address"],
+                            "ansible_host",
+                        )
+                    else:
+                        self.add_variable(
+                            device["name"], device["name"], "ansible_host"
+                        )
+                    if device["platform"]:
+                        self.add_variable(
+                            device["name"],
+                            ANSIBLE_LIB_MAPPER_REVERSE.get(
+                                NAPALM_LIB_MAPPER.get(
+                                    device["platform"]["napalm_driver"]
+                                )
+                            ),
```

Comment:
After the other feedback is collected as well, let's add some comments on what is going on here.

Reply:
Fixed